In [159]:
#Importing Pandas using an alias to avoid longer text


import pandas as pd

# Reading Data

There are several ways to read data by using Pandas which are specific for specific types of data formats.

Some of the most common and widely used functions in pandas are:-

1. read_csv :- It is used to read data present in CSV (Comma Seperated Values)

2. read_html :- This are used to read data present in form of tables in webpages

3. read_excel :- This is used to read the data present in excel sheets

In [160]:
#Let us read the data which is in csv format now.

draft1 = pd.read_csv('pokemon_data.csv')

draft1

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

There are two ways in which we can read the data in Pandas by using the read function.

Method 1. If the notebook we are working on is in the same directory as that of the dataset, we can directly read the dataset without    specifying the path

Method 2. If the notebook is present in a different directory, we can use the entire file path to read the dataset. Let us see           how to    do this below. Note that the default path has '/' as the seperators so we can use raw string to read or we           can change '/' to    '\' 

In [161]:
#This is the example for the method 2 mentioned above

example = pd.read_csv(r'C:\Users\Rahul M\Documents\Pandas\survey_results_public.csv')

example

Respondent                                         MainBranch Hobbyist  \
0               1             I am a student who is learning to code      Yes   
1               2             I am a student who is learning to code       No   
2               3  I am not primarily a developer, but I write co...      Yes   
3               4                     I am a developer by profession       No   
4               5                     I am a developer by profession      Yes   
...           ...                                                ...      ...   
88878       88377                                                NaN      Yes   
88879       88601                                                NaN       No   
88880       88802                                                NaN       No   
88881       88816                                                NaN       No   
88882       88863                                                NaN      Yes   

                                             OpenSourcer  \
0                                                  Never   
1                                Less than once per year   
2                                                  Never   
3                                                  Never   
4                             Once a month or more often   
...                                                  ...   
88878  Less than once a month but more than once per ...   
88879                                              Never   
88880                                              Never   
88881                                              Never   
88882                            Less than once per year   

                                              OpenSource  \
0      The quality of OSS and closed source software ...   
1      The quality of OSS and closed source software ...   
2      The quality of OSS and closed source software ...   
3      The quality of OSS and closed source software ...   
4      OSS is, on average, of HIGHER quality than pro...   
...                                                  ...   
88878  The quality of OSS and closed source software ...   
88879  The quality of OSS and closed source software ...   
88880                                                NaN   
88881  OSS is, on average, of HIGHER quality than pro...   
88882  OSS is, on average, of HIGHER quality than pro...   

                                              Employment  \
0                 Not employed, and not looking for work   
1                     Not employed, but looking for work   
2                                     Employed full-time   
3                                     Employed full-time   
4                                     Employed full-time   
...                                                  ...   
88878             Not employed, and not looking for work   
88879                                                NaN   
88880                                 Employed full-time   
88881  Independent contractor, freelancer, or self-em...   
88882             Not employed, and not looking for work   

                      Country         Student  \
0              United Kingdom              No   
1      Bosnia and Herzegovina  Yes, full-time   
2                    Thailand              No   
3               United States              No   
4                     Ukraine              No   
...                       ...             ...   
88878                  Canada              No   
88879                     NaN             NaN   
88880                     NaN             NaN   
88881                     NaN             NaN   
88882                   Spain  Yes, full-time   

                                                 EdLevel  \
0                              Primary/elementary school   
1      Secondary school (e.g. American high school, G...   
2               Bachelor’s degree (BA, BS, B.Eng., etc.)   
3               Bachelor’s degree (BA, BS,

# Head and Tail Function

Now that we have succesfully read the data, the first step would be to take a look at the data to understand the contents of it and get an idea of the basic information present in the data like Name of Columns, No. of Rows etc


This can be achieved by using the Head and Tail functions in Pandas

In [162]:
draft1.head()

#                   Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0  1              Bulbasaur  Grass  Poison  45      49       49       65   
1  2                Ivysaur  Grass  Poison  60      62       63       80   
2  3               Venusaur  Grass  Poison  80      82       83      100   
3  3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   
4  4             Charmander   Fire     NaN  39      52       43       60   

   Sp. Def  Speed  Generation  Legendary  
0       65     45           1      False  
1       80     60           1      False  
2      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False

As we can see, the head() function returns the top 5 rows present in the dataset. This is a useful function because we can quickly get an idea of the contents inside the dataset

Similarly, tail() function returns the last 5 rows by default

In [163]:
draft1.tail(10)

#                 Name   Type 1  Type 2   HP  Attack  Defense  Sp. Atk  \
790  714               Noibat   Flying  Dragon   40      30       35       45   
791  715              Noivern   Flying  Dragon   85      70       80       97   
792  716              Xerneas    Fairy     NaN  126     131       95      131   
793  717              Yveltal     Dark  Flying  126     131       95      131   
794  718     Zygarde50% Forme   Dragon  Ground  108     100      121       81   
795  719              Diancie     Rock   Fairy   50     100      150      100   
796  719  DiancieMega Diancie     Rock   Fairy   50     160      110      160   
797  720  HoopaHoopa Confined  Psychic   Ghost   80     110       60      150   
798  720   HoopaHoopa Unbound  Psychic    Dark   80     160       60      170   
799  721            Volcanion     Fire   Water   80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
790       40     55           6      False  
791       80    123           6      False  
792       98     99           6       True  
793       98     99           6       True  
794       95     95           6       True  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True

If the value is not specified, head() and tail() returns the first and the last 5 rows respecitvely but if the values are specified, they return the suitable number of rows accordingly as shown above

There are many parameters which are available inside read_csv() so lets look at a few of them right now

1. header = 'infer'
2. names = None
3. index_col = None



## header = 'infer' and names = None

By default, when we read any dataset, the parameter header is set to be 'infer'. Infer stands for inferring the header names from the dataset itself.

If no names are passed as column names in the read_csv() function, header becomes equal to 'infer' and Pandas automatically pass in the first row elements as column names by default. In this case, names are set to None

If names are passed explicitly, header becomes equal to none and those names are used as Column Names.

Let us see this with the help of examples

In [164]:
draft2 = pd.read_csv('pokemon_data.csv')

Here, the names of the columns are not explicitly mentioned, so header is equal to infer and column names are the first row values of the dataset

In [165]:
colnames = list(draft1.columns)
colnames.append('New Name')


draft3 = pd.read_csv('pokemon_data.csv',names=colnames)


In [166]:
draft3

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      #                   Name   Type 1  Type 2  HP  Attack  Defense   
1      1              Bulbasaur    Grass  Poison  45      49       49   
2      2                Ivysaur    Grass  Poison  60      62       63   
3      3               Venusaur    Grass  Poison  80      82       83   
4      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
..   ...                    ...      ...     ...  ..     ...      ...   
796  719                Diancie     Rock   Fairy  50     100      150   
797  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
798  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
799  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
800  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  New Name  
0    Sp. Atk  Sp. Def  Speed  Generation  Legendary       NaN  
1         65       65     45           1      FALSE       NaN  
2         80       80     60           1      FALSE       NaN  
3        100      100     80           1      FALSE       NaN  
4        122      120     80           1      FALSE       NaN  
..       ...      ...    ...         ...        ...       ...  
796      100      150     50           6       TRUE       NaN  
797      160      110    110           6       TRUE       NaN  
798      150      130     70           6       TRUE       NaN  
799      170      130     80           6       TRUE       NaN  
800      130       90     70           6       TRUE       NaN  

[801 rows x 13 columns]

Here, the column names are explicitly mentioned and passed in the read_csv() function so this time header becomes equal to None and these names are used as column names. 

Notice that the usual column names occupied the first row this time

## index_col

As the name implies, index_col refers to the index column of the dataframe. By default, index_col is set to None and indexing usually starts from 0 in this case till the end of number of rows.

There are two ways in which we can index the dataframe in Pandas

1. Single Indexing
2. Multi Indexing

In [167]:
draft1.head()

#                   Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0  1              Bulbasaur  Grass  Poison  45      49       49       65   
1  2                Ivysaur  Grass  Poison  60      62       63       80   
2  3               Venusaur  Grass  Poison  80      82       83      100   
3  3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   
4  4             Charmander   Fire     NaN  39      52       43       60   

   Sp. Def  Speed  Generation  Legendary  
0       65     45           1      False  
1       80     60           1      False  
2      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False

In [168]:
ex0 = pd.read_csv('pokemon_data.csv',index_col='#')

ex0.head()

Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
#                                                                       
1              Bulbasaur  Grass  Poison  45      49       49       65   
2                Ivysaur  Grass  Poison  60      62       63       80   
3               Venusaur  Grass  Poison  80      82       83      100   
3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   
4             Charmander   Fire     NaN  39      52       43       60   

   Sp. Def  Speed  Generation  Legendary  
#                                         
1       65     45           1      False  
2       80     60           1      False  
3      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False

Notice the differences in draft1 and ex0 dataframes.

In draft1, we did not specify the index column so the indexing started from 0 and continues onwards till the end.

But in ex0, since we explicitly mentioned index_col = '#' column, the '#' column is set as index so in this case, indexing started from 1.

This is an example of Single Indexing

In [169]:
ex1 = pd.read_csv('pokemon_data.csv',index_col = ['#','Legendary'])

ex1.head(20)

Name Type 1  Type 2  HP  Attack  Defense  \
#  Legendary                                                                  
1  False                      Bulbasaur  Grass  Poison  45      49       49   
2  False                        Ivysaur  Grass  Poison  60      62       63   
3  False                       Venusaur  Grass  Poison  80      82       83   
   False          VenusaurMega Venusaur  Grass  Poison  80     100      123   
4  False                     Charmander   Fire     NaN  39      52       43   
5  False                     Charmeleon   Fire     NaN  58      64       58   
6  False                      Charizard   Fire  Flying  78      84       78   
   False      CharizardMega Charizard X   Fire  Dragon  78     130      111   
   False      CharizardMega Charizard Y   Fire  Flying  78     104       78   
7  False                       Squirtle  Water     NaN  44      48       65   
8  False                      Wartortle  Water     NaN  59      63       80   
9  False                      Blastoise  Water     NaN  79      83      100   
   False        BlastoiseMega Blastoise  Water     NaN  79     103      120   
10 False                       Caterpie    Bug     NaN  45      30       35   
11 False                        Metapod    Bug     NaN  50      20       55   
12 False                     Butterfree    Bug  Flying  60      45       50   
13 False                         Weedle    Bug  Poison  40      35       30   
14 False                         Kakuna    Bug  Poison  45      25       50   
15 False                       Beedrill    Bug  Poison  65      90       40   
   False          BeedrillMega Beedrill    Bug  Poison  65     150       40   

              Sp. Atk  Sp. Def  Speed  Generation  
#  Legendary                                       
1  False           65       65     45           1  
2  False           80       80     60           1  
3  False          100      100     80           1  
   False          122      120     80           1  
4  False           60       50     65           1  
5  False           80       65     80           1  
6  False          109       85    100           1  
   False          130       85    100           1  
   False          159      115    100           1  
7  False           50       64     43           1  
8  False           65       80     58           1  
9  False           85      105     78           1  
   False          135      115     78           1  
10 False           20       20     45           1  
11 False           25       25     30           1  
12 False           90       80     70           1  
13 False           20       20     50           1  
14 False           25       25     35           1  
15 False           45       80     75           1  
   False           15       80    145           1

ex2 dataframe is an example of Multi Indexing wherein the dataframe has two indexes namely '#' and 'Legendary'

These are the various types of indexing we can use in Pandas

# isnull() function

The isnull() function is one of the most used functions while working with Pandas Dataframes. Often the data we obtain from the real-world is messy and it has a lot of NaN or Null Values that comes along with it. 

The isnull() function returns the Dataframe consisting of Trues and Falses based on if the data present in that location is null or not.

Let us see the example

In [170]:
draft1.isnull()

#   Name  Type 1  Type 2     HP  Attack  Defense  Sp. Atk  Sp. Def  \
0    False  False   False   False  False   False    False    False    False   
1    False  False   False   False  False   False    False    False    False   
2    False  False   False   False  False   False    False    False    False   
3    False  False   False   False  False   False    False    False    False   
4    False  False   False    True  False   False    False    False    False   
..     ...    ...     ...     ...    ...     ...      ...      ...      ...   
795  False  False   False   False  False   False    False    False    False   
796  False  False   False   False  False   False    False    False    False   
797  False  False   False   False  False   False    False    False    False   
798  False  False   False   False  False   False    False    False    False   
799  False  False   False   False  False   False    False    False    False   

     Speed  Generation  Legendary  
0    False       False      False  
1    False       False      False  
2    False       False      False  
3    False       False      False  
4    False       False      False  
..     ...         ...        ...  
795  False       False      False  
796  False       False      False  
797  False       False      False  
798  False       False      False  
799  False       False      False  

[800 rows x 12 columns]

Just by looking at this dataframe, we can see a clear shortcoming. Although this does return the true or falses, since the data is huge, we cannot clearly see how many columns have null values

To overcome this problem, we use sum() function which returns the number of null values in each column of the dataframe

In [171]:
draft1.isnull().sum()

#               0
Name            0
Type 1          0
Type 2        386
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

As we can see, Type 2 Column in this dataframe has 386 null values.

## Excel Reading Syntax

In order to read the data from excel sheets, we follow a usual syntax which looks like this.

pd.read_excel('name.xlsx',sheet_name='name')

Since one excel file can have multiple sheets, we use the sheet name to read the data from specific sheet

In [172]:
excel1 = pd.read_excel(r'C:\Users\Rahul M\Documents\Book1.xlsx',sheet_name='Details')

excel1

Name  Age  Salary
0  James   23   12000
1  Jason   21   23000
2    Tim   24   30000

## Reading from URL

Let us read the data from a url which contains the data in the form of tables and not as raw data. Inorder to read the data from such sources, we use read_html() function to return the table of data


In [173]:
url = 'https://www.basketball-reference.com/leagues/NBA_2015_totals.html'

exa1 = pd.read_html(url)

exa1

[      Rk          Player Pos Age   Tm   G  GS    MP   FG  FGA  ...   FT%  ORB  \
 0      1      Quincy Acy  PF  24  NYK  68  22  1287  152  331  ...  .784   79   
 1      2    Jordan Adams  SG  20  MEM  30   0   248   35   86  ...  .609    9   
 2      3    Steven Adams   C  21  OKC  70  67  1771  217  399  ...  .502  199   
 3      4     Jeff Adrien  PF  28  MIN  17   0   215   19   44  ...  .579   23   
 4      5   Arron Afflalo  SG  29  TOT  78  72  2502  375  884  ...  .843   27   
 ..   ...             ...  ..  ..  ...  ..  ..   ...  ...  ...  ...   ...  ...   
 670  490  Thaddeus Young  PF  26  TOT  76  68  2434  451  968  ...  .655  127   
 671  490  Thaddeus Young  PF  26  MIN  48  48  1605  289  641  ...  .682   75   
 672  490  Thaddeus Young  PF  26  BRK  28  20   829  162  327  ...  .606   52   
 673  491     Cody Zeller   C  22  CHO  62  45  1487  172  373  ...  .774   97   
 674  492    Tyler Zeller   C  25  BOS  82  59  1731  340  619  ...  .823  146   
 
      DRB  TRB

In [174]:
type(exa1)

list

Even though the code worked properly, we can see that the data is not present in the form of a dataframe, instead it is in the form of a list. 

This is because of the fact that, when we use read_html() function on a url, it returns the data which are present in tables. These tables can be one or many depending on the website.

But, since it is difficult to display all the tables at once, this function appends all the dataframes in a list, essentially wrapping them.

So inorder to access individual dataframes, we need to index the list

In [175]:
dfa = exa1[0]
dfa

Rk          Player Pos Age   Tm   G  GS    MP   FG  FGA  ...   FT%  ORB  \
0      1      Quincy Acy  PF  24  NYK  68  22  1287  152  331  ...  .784   79   
1      2    Jordan Adams  SG  20  MEM  30   0   248   35   86  ...  .609    9   
2      3    Steven Adams   C  21  OKC  70  67  1771  217  399  ...  .502  199   
3      4     Jeff Adrien  PF  28  MIN  17   0   215   19   44  ...  .579   23   
4      5   Arron Afflalo  SG  29  TOT  78  72  2502  375  884  ...  .843   27   
..   ...             ...  ..  ..  ...  ..  ..   ...  ...  ...  ...   ...  ...   
670  490  Thaddeus Young  PF  26  TOT  76  68  2434  451  968  ...  .655  127   
671  490  Thaddeus Young  PF  26  MIN  48  48  1605  289  641  ...  .682   75   
672  490  Thaddeus Young  PF  26  BRK  28  20   829  162  327  ...  .606   52   
673  491     Cody Zeller   C  22  CHO  62  45  1487  172  373  ...  .774   97   
674  492    Tyler Zeller   C  25  BOS  82  59  1731  340  619  ...  .823  146   

     DRB  TRB  AST  STL BLK  TOV   PF   PTS  
0    222  301   68   27  22   60  147   398  
1     19   28   16   16   7   14   24    94  
2    324  523   66   38  86   99  222   537  
3     54   77   15    4   9    9   30    60  
4    220  247  129   41   7  116  167  1035  
..   ...  ...  ...  ...  ..  ...  ...   ...  
670  284  411  173  124  25  117  171  1071  
671  170  245  135   86  17   75  115   685  
672  114  166   38   38   8   42   56   386  
673  265  362  100   34  49   62  156   472  
674  319  465  113   18  52   76  205   833  

[675 rows x 30 columns]

Now we can see that this data is a dataframe

# Selecting one or multiple columns

In Pandas, we can select one or multiple columns by passing their names in a list within the dataframe.

Lets look at how we do it

In [176]:
draft1['Name']

0                  Bulbasaur
1                    Ivysaur
2                   Venusaur
3      VenusaurMega Venusaur
4                 Charmander
               ...          
795                  Diancie
796      DiancieMega Diancie
797      HoopaHoopa Confined
798       HoopaHoopa Unbound
799                Volcanion
Name: Name, Length: 800, dtype: object

In [177]:
type(draft1['Name'])

pandas.core.series.Series

In [178]:
draft1[['Name']]

Name
0                Bulbasaur
1                  Ivysaur
2                 Venusaur
3    VenusaurMega Venusaur
4               Charmander
..                     ...
795                Diancie
796    DiancieMega Diancie
797    HoopaHoopa Confined
798     HoopaHoopa Unbound
799              Volcanion

[800 rows x 1 columns]

In [179]:
type(draft1[['Name']])

pandas.core.frame.DataFrame

As we can notice from the past 4 cells, if we mention the name of the column within a single [], pandas interpret this data as Pandas Series and not Pandas Dataframe. Inorder for us to make this into a dataframe, we have to pass it as a list with the [].

The same is the case for multiple columns selecting. If we pass multiple column names without a list, we will encounter an error since a Pandas Series can only have one column

In [180]:
draft1[['Name','Type 1']]

Name   Type 1
0                Bulbasaur    Grass
1                  Ivysaur    Grass
2                 Venusaur    Grass
3    VenusaurMega Venusaur    Grass
4               Charmander     Fire
..                     ...      ...
795                Diancie     Rock
796    DiancieMega Diancie     Rock
797    HoopaHoopa Confined  Psychic
798     HoopaHoopa Unbound  Psychic
799              Volcanion     Fire

[800 rows x 2 columns]

## .columns

Inorder to quickly look at the number of columns in a dataframe, we can use the .columns function.

This function returns all the column names of the dataframe in a list format

In [181]:
draft1.columns

Index(['#', 'Name', 'Type 1', 'Type 2', 'HP', 'Attack', 'Defense', 'Sp. Atk',
       'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

## .dtypes

We can navigate through the datatypes of all the columns by using the .dtypes function in Pandas.

In [182]:
draft1.dtypes

#              int64
Name          object
Type 1        object
Type 2        object
HP             int64
Attack         int64
Defense        int64
Sp. Atk        int64
Sp. Def        int64
Speed          int64
Generation     int64
Legendary       bool
dtype: object

## nrows parameter

In [184]:
draft1.head(30)

#                   Name  Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0    1              Bulbasaur   Grass  Poison  45      49       49       65   
1    2                Ivysaur   Grass  Poison  60      62       63       80   
2    3               Venusaur   Grass  Poison  80      82       83      100   
3    3  VenusaurMega Venusaur   Grass  Poison  80     100      123      122   
4    4             Charmander    Fire     NaN  39      52       43       60   
..  ..                    ...     ...     ...  ..     ...      ...      ...   
25  20               Raticate  Normal     NaN  55      81       60       50   
26  21                Spearow  Normal  Flying  40      60       30       31   
27  22                 Fearow  Normal  Flying  65      90       65       61   
28  23                  Ekans  Poison     NaN  35      60       44       40   
29  24                  Arbok  Poison     NaN  60      85       69       65   

    Sp. Def  Speed  Generation  Legendary  
0        65     45           1      False  
1        80     60           1      False  
2       100     80           1      False  
3       120     80           1      False  
4        50     65           1      False  
..      ...    ...         ...        ...  
25       70     97           1      False  
26       31     70           1      False  
27       61    100           1      False  
28       54     55           1      False  
29       79     80           1      False  

[30 rows x 12 columns]

In [185]:
draftx = pd.read_csv('pokemon_data.csv',nrows = 30)

draftx

#                   Name  Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0    1              Bulbasaur   Grass  Poison  45      49       49       65   
1    2                Ivysaur   Grass  Poison  60      62       63       80   
2    3               Venusaur   Grass  Poison  80      82       83      100   
3    3  VenusaurMega Venusaur   Grass  Poison  80     100      123      122   
4    4             Charmander    Fire     NaN  39      52       43       60   
..  ..                    ...     ...     ...  ..     ...      ...      ...   
25  20               Raticate  Normal     NaN  55      81       60       50   
26  21                Spearow  Normal  Flying  40      60       30       31   
27  22                 Fearow  Normal  Flying  65      90       65       61   
28  23                  Ekans  Poison     NaN  35      60       44       40   
29  24                  Arbok  Poison     NaN  60      85       69       65   

    Sp. Def  Speed  Generation  Legendary  
0        65     45           1      False  
1        80     60           1      False  
2       100     80           1      False  
3       120     80           1      False  
4        50     65           1      False  
..      ...    ...         ...        ...  
25       70     97           1      False  
26       31     70           1      False  
27       61    100           1      False  
28       54     55           1      False  
29       79     80           1      False  

[30 rows x 12 columns]

Just by first look, we can see that both head() function and nrows parameter returns the same result but nrows has another use which we will see now.

* When we use head() function and pass 30 as a parameter, the "entire" dataframe gets readed and first 30 rows are displayed.

* But when we use the parameter nrows = 30, only the first 30 rows are read not the entire data and gets displayed.

Uses of nrows:-

While dealing with a large dataset consisting of GBs of data, if we try to use head(), the computational speed reduces. But if we use nrows and just display the necessary rows, the task can easily be done without wasting RAM.

## chunksize parameter

The chunksize parameter is similar to the nrows wherein the entire data is not usually read. 

But the difference lies in the size of the data in bytes. By setting chunksize as x bytes, the data is read till it reaches the x bytes and is stopped if it immediately exceeds the specified size.

These are also useful when dealing with larger datasets where the size is known.

Lets see an example

In [186]:
chunk = pd.read_csv('pokemon_data.csv',chunksize=100)

#When we use chunksize, a text file is returned so we iterate through it to display data

for i in chunk:
    print(i)

     #                   Name  Type 1  Type 2   HP  Attack  Defense  Sp. Atk  \
0    1              Bulbasaur   Grass  Poison   45      49       49       65   
1    2                Ivysaur   Grass  Poison   60      62       63       80   
2    3               Venusaur   Grass  Poison   80      82       83      100   
3    3  VenusaurMega Venusaur   Grass  Poison   80     100      123      122   
4    4             Charmander    Fire     NaN   39      52       43       60   
..  ..                    ...     ...     ...  ...     ...      ...      ...   
95  88                 Grimer  Poison     NaN   80      80       50       40   
96  89                    Muk  Poison     NaN  105     105       75       65   
97  90               Shellder   Water     NaN   30      65      100       45   
98  91               Cloyster   Water     Ice   50      95      180       85   
99  92                 Gastly   Ghost  Poison   30      35       30      100   

    Sp. Def  Speed  Generation  Legenda

## Writing Data to a Text Format

Till now, we have seen how to read data from a dataset and display it as a dataframe in pandas.

But in real life, once we finish up editing and modifying the dataset, we would sometimes need to save it as a file in our local storage to use it another time.

This task can be accomplished in pandas by using to_csv() function.

In [187]:
#Now, lets see how to write the data into a csv format.

c1 = pd.read_csv('pokemon_data.csv',nrows = 100)

c1

#                   Name  Type 1  Type 2   HP  Attack  Defense  Sp. Atk  \
0    1              Bulbasaur   Grass  Poison   45      49       49       65   
1    2                Ivysaur   Grass  Poison   60      62       63       80   
2    3               Venusaur   Grass  Poison   80      82       83      100   
3    3  VenusaurMega Venusaur   Grass  Poison   80     100      123      122   
4    4             Charmander    Fire     NaN   39      52       43       60   
..  ..                    ...     ...     ...  ...     ...      ...      ...   
95  88                 Grimer  Poison     NaN   80      80       50       40   
96  89                    Muk  Poison     NaN  105     105       75       65   
97  90               Shellder   Water     NaN   30      65      100       45   
98  91               Cloyster   Water     Ice   50      95      180       85   
99  92                 Gastly   Ghost  Poison   30      35       30      100   

    Sp. Def  Speed  Generation  Legendary  
0        65     45           1      False  
1        80     60           1      False  
2       100     80           1      False  
3       120     80           1      False  
4        50     65           1      False  
..      ...    ...         ...        ...  
95       50     25           1      False  
96      100     50           1      False  
97       25     40           1      False  
98       45     70           1      False  
99       35     80           1      False  

[100 rows x 12 columns]

In [188]:
c1.to_csv('Compressed_Pokemon_Data.csv')

#Now, a new csv file appears with the name Compressed_Pokemon_Data which has only 100 entries

If we want to save the dataset into another directory, we can specify the complete path inplace of only the name and get the job done, just like with reading data from another directory


Let us look at a few more scenarios and try to solve them.

1. Sometimes, we would like to store the data by using a different seperator and not just commas.

2. In the case of Online Machine Learning Competitions, usually we have to upload the predictions data without any header or index for the algorithm to scan and display our score.

For the scenario one, we can use the parameter sep = '@' to get our job done. This is similar to the syntax of sep in read_csv() function as well.

Now, lets do this operation and look at the resulting dataset

In [189]:
c1.to_csv('Compressed_Pokemon_Data.csv',sep='@')

dfa = pd.read_csv('Compressed_Pokemon_Data.csv')

dfa

@#@Name@Type 1@Type 2@HP@Attack@Defense@Sp. Atk@Sp. Def@Speed@Generation@Legendary
0   0@1@Bulbasaur@Grass@Poison@45@49@49@65@65@45@1...                                
1   1@2@Ivysaur@Grass@Poison@60@62@63@80@80@60@1@F...                                
2   2@3@Venusaur@Grass@Poison@80@82@83@100@100@80@...                                
3   3@3@VenusaurMega Venusaur@Grass@Poison@80@100@...                                
4      4@4@Charmander@Fire@@39@52@43@60@50@65@1@False                                
..                                                ...                                
95     95@88@Grimer@Poison@@80@80@50@40@50@25@1@False                                
96     96@89@Muk@Poison@@105@105@75@65@100@50@1@False                                
97   97@90@Shellder@Water@@30@65@100@45@25@40@1@False                                
98  98@91@Cloyster@Water@Ice@50@95@180@85@45@70@1@...                                
99  99@92@Gastly@Ghost@Poison@30@35@30@100@35@80@1...                                

[100 rows x 1 columns]

As we can see, the data is modified and @ is the new sepearator now.

In scenario 2, we don't want the data to have an index or column names while writing. We can accomplish this task by using two parameters called 'header' and 'index'.

The default value of 'header' and 'index' are set as True so we need to explicitly change them inorder to not get column names and index.

Now, lets code this

In [190]:
c1 = pd.read_csv('pokemon_data.csv',nrows = 100)

c1.to_csv('Compressed_Data.csv',header=False,index=False)

c11 = pd.read_csv('Compressed_Data.csv')

c11.head()

1              Bulbasaur  Grass  Poison  45   49  49.1   65  65.1  45.1  \
0  2                Ivysaur  Grass  Poison  60   62    63   80    80    60   
1  3               Venusaur  Grass  Poison  80   82    83  100   100    80   
2  3  VenusaurMega Venusaur  Grass  Poison  80  100   123  122   120    80   
3  4             Charmander   Fire     NaN  39   52    43   60    50    65   
4  5             Charmeleon   Fire     NaN  58   64    58   80    65    80   

   1.1  False  
0    1  False  
1    1  False  
2    1  False  
3    1  False  
4    1  False

Now, we can clearly see that both the column names and the index are missing when we read the dataset.

Now, there might be cases where we want to view the data first before writing it to a file name to see if everything worked out well, rather than writing it and then going to the folder to see.

This can be done by using a default library in python called 'sys' 

In [191]:
import sys

draft1.to_csv(sys.stdout,na_rep='NO VALUE')

#na_rep replaces the nan values with a default value in this case NO VALUE. sys.stdout displayes the text in clipboard

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
4,4,Charmander,Fire,NO VALUE,39,52,43,60,50,65,1,False
5,5,Charmeleon,Fire,NO VALUE,58,64,58,80,65,80,1,False
6,6,Charizard,Fire,Flying,78,84,78,109,85,100,1,False
7,6,CharizardMega Charizard X,Fire,Dragon,78,130,111,130,85,100,1,False
8,6,CharizardMega Charizard Y,Fire,Flying,78,104,78,159,115,100,1,False
9,7,Squirtle,Water,NO VALUE,44,48,65,50,64,43,1,False
10,8,Wartortle,Water,NO VALUE,59,63,80,65,80,58,1,False
11,9,Blastoise,Water,NO VALUE,79,83,100,85,105,78,1,False
12,9,BlastoiseMega Blastoise,Water,NO VALUE,79,103,120,135,115,78,1,False
13,10,Caterpie,Bug,NO VALUE,45,30,35,20,20,45,1,False
14,11,Metapod,Bug,NO VALUE,50,20,55,25,25,30,1,False
15,12,Butterfree,Bug,Flying,60,45

By using the function 'sys.stdout', we can print the data in the clipboard and check it before writing it to a file

We can also use the header and index parameters combined with 'sys.stdout' which are done in the next few cells

In [192]:
c1.to_csv(sys.stdout,header=False,index=False)


#We can see column names and index values are missing as we wanted

1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
4,Charmander,Fire,,39,52,43,60,50,65,1,False
5,Charmeleon,Fire,,58,64,58,80,65,80,1,False
6,Charizard,Fire,Flying,78,84,78,109,85,100,1,False
6,CharizardMega Charizard X,Fire,Dragon,78,130,111,130,85,100,1,False
6,CharizardMega Charizard Y,Fire,Flying,78,104,78,159,115,100,1,False
7,Squirtle,Water,,44,48,65,50,64,43,1,False
8,Wartortle,Water,,59,63,80,65,80,58,1,False
9,Blastoise,Water,,79,83,100,85,105,78,1,False
9,BlastoiseMega Blastoise,Water,,79,103,120,135,115,78,1,False
10,Caterpie,Bug,,45,30,35,20,20,45,1,False
11,Metapod,Bug,,50,20,55,25,25,30,1,False
12,Butterfree,Bug,Flying,60,45,50,90,80,70,1,False
13,Weedle,Bug,Poison,40,35,30,20,20,50,1,False
14,Kakuna,Bug,Poison,45,25,50,25,25,35,1,False
15,Beedrill,Bug,Poison,65,90,40,45,80,75,1,False
15,BeedrillMega Beedr

## columns parameter

By using the 'columns' parameter, we can select and filter only those columns which we want to write to the file name and not all the columns.

This can come handy in online competitions where we want only the prediciton column to be in the file and not all columns.

This is how we do it

In [193]:
#We can even filter and select specific  column names as below

c1.to_csv(sys.stdout,columns = ['Name','Generation'],index=False)




Name,Generation
Bulbasaur,1
Ivysaur,1
Venusaur,1
VenusaurMega Venusaur,1
Charmander,1
Charmeleon,1
Charizard,1
CharizardMega Charizard X,1
CharizardMega Charizard Y,1
Squirtle,1
Wartortle,1
Blastoise,1
BlastoiseMega Blastoise,1
Caterpie,1
Metapod,1
Butterfree,1
Weedle,1
Kakuna,1
Beedrill,1
BeedrillMega Beedrill,1
Pidgey,1
Pidgeotto,1
Pidgeot,1
PidgeotMega Pidgeot,1
Rattata,1
Raticate,1
Spearow,1
Fearow,1
Ekans,1
Arbok,1
Pikachu,1
Raichu,1
Sandshrew,1
Sandslash,1
Nidoran (Female),1
Nidorina,1
Nidoqueen,1
Nidoran (Male),1
Nidorino,1
Nidoking,1
Clefairy,1
Clefable,1
Vulpix,1
Ninetales,1
Jigglypuff,1
Wigglytuff,1
Zubat,1
Golbat,1
Oddish,1
Gloom,1
Vileplume,1
Paras,1
Parasect,1
Venonat,1
Venomoth,1
Diglett,1
Dugtrio,1
Meowth,1
Persian,1
Psyduck,1
Golduck,1
Mankey,1
Primeape,1
Growlithe,1
Arcanine,1
Poliwag,1
Poliwhirl,1
Poliwrath,1
Abra,1
Kadabra,1
Alakazam,1
AlakazamMega Alakazam,1
Machop,1
Machoke,1
Machamp,1
Bellsprout,1
Weepinbell,1
Victreebel,1
Tentacool,1
Tentacruel,1
Geodude,1
Gravel

## date_range()

Let us use another function available in pandas which is called date_range() function.

The date_range() is similar to the function arange() present in Numpy. This function() when used, returns a list of dates which we have specified inside its arguments.

Its syntax is like :- date_range(start,end,periods)

Here start and stop defines the start dates and the stop dates and periods refers to the number of dates which we need to return.

In [194]:
dates = pd.date_range('1/1/2000',periods=7)
ts= pd.Series(np.arange(7),index=dates)
ts

2000-01-01    0
2000-01-02    1
2000-01-03    2
2000-01-04    3
2000-01-05    4
2000-01-06    5
2000-01-07    6
Freq: D, dtype: int32

in the above code, we first created a list of dates by using the date_range() function and used it as an index for a Pandas Series of numbers from 1 to 6.

In [195]:
type(dates)

pandas.core.indexes.datetimes.DatetimeIndex

# Reading Data without using Pandas

Now, let us look at few of the ways in which we can read the data without using Pandas. Although these are NOT recommended, it is good to know the procedure to do them

Here, we use another library available in python named 'csv' for reading the data. The process is as follows

1. Importing the csv library
2. By using the open function, we try to open and store all the contents of the file in a variable.
3. This variable is accessed and the contents are read into another variable by using reader() function in csv library
4. This reader object is iterated through to unpack the entire data

In [196]:
import csv
f = open('pokemon_data.csv')

reader = csv.reader(f)
reader

In [197]:
for i in reader:
    print(i)

['#', 'Name', 'Type 1', 'Type 2', 'HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary']
['1', 'Bulbasaur', 'Grass', 'Poison', '45', '49', '49', '65', '65', '45', '1', 'FALSE']
['2', 'Ivysaur', 'Grass', 'Poison', '60', '62', '63', '80', '80', '60', '1', 'FALSE']
['3', 'Venusaur', 'Grass', 'Poison', '80', '82', '83', '100', '100', '80', '1', 'FALSE']
['3', 'VenusaurMega Venusaur', 'Grass', 'Poison', '80', '100', '123', '122', '120', '80', '1', 'FALSE']
['4', 'Charmander', 'Fire', '', '39', '52', '43', '60', '50', '65', '1', 'FALSE']
['5', 'Charmeleon', 'Fire', '', '58', '64', '58', '80', '65', '80', '1', 'FALSE']
['6', 'Charizard', 'Fire', 'Flying', '78', '84', '78', '109', '85', '100', '1', 'FALSE']
['6', 'CharizardMega Charizard X', 'Fire', 'Dragon', '78', '130', '111', '130', '85', '100', '1', 'FALSE']
['6', 'CharizardMega Charizard Y', 'Fire', 'Flying', '78', '104', '78', '159', '115', '100', '1', 'FALSE']
['7', 'Squirtle', 'Water', '', '44', '48', '65'

By using this method, we can read the data in the form of a collection but we cant easily manipulate the data since they are not in a dataframe

# JSON DATA

JSON stands for "Java Script Object Notation".

JSON is also another popular type of formats to represent data so it is important to know how to read and manipulate them by using Pandas.

Typically, JSON data looks like a complex connections of data within data in the form of dictionaries and lists.

Let us look at an example below

In [123]:
obj = '''
{
	"id": "0001",
	"type": "donut",
	"name": "Cake",
	"ppu": 0.55,
	"batters":
		{
			"batter":
				[
					{ "id": "1001", "type": "Regular" },
					{ "id": "1002", "type": "Chocolate" },
					{ "id": "1003", "type": "Blueberry" },
					{ "id": "1004", "type": "Devil's Food" }
				]
		},
	"topping":
		[
			{ "id": "5001", "type": "None" },
			{ "id": "5002", "type": "Glazed" },
			{ "id": "5005", "type": "Sugar" },
			{ "id": "5007", "type": "Powdered Sugar" },
			{ "id": "5006", "type": "Chocolate with Sprinkles" },
			{ "id": "5003", "type": "Chocolate" },
			{ "id": "5004", "type": "Maple" }
		]
}
'''
type(obj)

str

The data inside the ''' is the JSON data and in this case, we are intialising it as a string to work with Pandas.

The default way is to save the data in .json format and use a default function 'read_json()' to read the json directly. But for now, let us look at how we can work with JSON data in strings.

In [124]:
import json
result = json.loads(obj)
result

{'id': '0001',
 'type': 'donut',
 'name': 'Cake',
 'ppu': 0.55,
 'batters': {'batter': [{'id': '1001', 'type': 'Regular'},
   {'id': '1002', 'type': 'Chocolate'},
   {'id': '1003', 'type': 'Blueberry'},
   {'id': '1004', 'type': "Devil's Food"}]},
 'topping': [{'id': '5001', 'type': 'None'},
  {'id': '5002', 'type': 'Glazed'},
  {'id': '5005', 'type': 'Sugar'},
  {'id': '5007', 'type': 'Powdered Sugar'},
  {'id': '5006', 'type': 'Chocolate with Sprinkles'},
  {'id': '5003', 'type': 'Chocolate'},
  {'id': '5004', 'type': 'Maple'}]}

Here, we used the library 'json' inorder to display the json data.

The .loads() method in json takes in the string of json data as an input and returns the dictionary as an output

In [198]:
type(result)

dict

In order to convert the dictionary data into string format, we can use the .dumps() method and pass in the dictionary data as an input. This method is shown in the below cell

In [200]:
backtostr = json.dumps(result)
type(backtostr)

str

Now that we have successfully converted the JSON data from string to dictionary, we can use a portion of the data or the entire data to create a datadrame in Pandas.

Additionally we can specify only those column names which we need to further filter the data

In [203]:
topping = pd.DataFrame(result['topping'],columns=['id','type'])
topping

id                      type
0  5001                      None
1  5002                    Glazed
2  5005                     Sugar
3  5007            Powdered Sugar
4  5006  Chocolate with Sprinkles
5  5003                 Chocolate
6  5004                     Maple

In [130]:
result['topping']

[{'id': '5001', 'type': 'None'},
 {'id': '5002', 'type': 'Glazed'},
 {'id': '5005', 'type': 'Sugar'},
 {'id': '5007', 'type': 'Powdered Sugar'},
 {'id': '5006', 'type': 'Chocolate with Sprinkles'},
 {'id': '5003', 'type': 'Chocolate'},
 {'id': '5004', 'type': 'Maple'}]

We can index even a dictionary inside a dictionary by specifying the appropriate index slicing as shown below.

In this example, we created a dataframe of batter dictionary which is available inside batters dictionary which is a part of the result dictionary

In [204]:
pd.DataFrame(result['batters']['batter'],columns=['id','type'])

id          type
0  1001       Regular
1  1002     Chocolate
2  1003     Blueberry
3  1004  Devil's Food

For reading a default JSON file, we use the read_json() function as shown below

In [208]:
df = pd.read_json('https://api.github.com/repos/pandas-dev/pandas/issues')

df.head(5)

url  \
0  https://api.github.com/repos/pandas-dev/pandas...   
1  https://api.github.com/repos/pandas-dev/pandas...   
2  https://api.github.com/repos/pandas-dev/pandas...   
3  https://api.github.com/repos/pandas-dev/pandas...   
4  https://api.github.com/repos/pandas-dev/pandas...   

                                   repository_url  \
0  https://api.github.com/repos/pandas-dev/pandas   
1  https://api.github.com/repos/pandas-dev/pandas   
2  https://api.github.com/repos/pandas-dev/pandas   
3  https://api.github.com/repos/pandas-dev/pandas   
4  https://api.github.com/repos/pandas-dev/pandas   

                                          labels_url  \
0  https://api.github.com/repos/pandas-dev/pandas...   
1  https://api.github.com/repos/pandas-dev/pandas...   
2  https://api.github.com/repos/pandas-dev/pandas...   
3  https://api.github.com/repos/pandas-dev/pandas...   
4  https://api.github.com/repos/pandas-dev/pandas...   

                                        comments_url  \
0  https://api.github.com/repos/pandas-dev/pandas...   
1  https://api.github.com/repos/pandas-dev/pandas...   
2  https://api.github.com/repos/pandas-dev/pandas...   
3  https://api.github.com/repos/pandas-dev/pandas...   
4  https://api.github.com/repos/pandas-dev/pandas...   

                                          events_url  \
0  https://api.github.com/repos/pandas-dev/pandas...   
1  https://api.github.com/repos/pandas-dev/pandas...   
2  https://api.github.com/repos/pandas-dev/pandas...   
3  https://api.github.com/repos/pandas-dev/pandas...   
4  https://api.github.com/repos/pandas-dev/pandas...   

                                            html_url         id  \
0  https://github.com/pandas-dev/pandas/issues/41180  868934967   
1    https://github.com/pandas-dev/pandas/pull/41179  868911996   
2    https://github.com/pandas-dev/pandas/pull/41178  868617307   
3    https://github.com/pandas-dev/pandas/pull/41177  868564716   
4    https://github.com/pandas-dev/pandas/pull/41176  868435746   

                            node_id  number  \
0          MDU6SXNzdWU4Njg5MzQ5Njc=   41180   
1  MDExOlB1bGxSZXF1ZXN0NjI0MzU4MzA2   41179   
2  MDExOlB1bGxSZXF1ZXN0NjI0MTEzNTIw   41178   
3  MDExOlB1bGxSZXF1ZXN0NjI0MDY5OTAw   41177   
4  MDExOlB1bGxSZXF1ZXN0NjIzOTYyOTA3   41176   

                                               title  ...  \
0     DOC: DataFrame on df.nunique causing confusion  ...   
1       TYP: cleanup typing in core.indexes.range.py  ...   
2  TST: run frame/series arithmetic tests with+wi...  ...   
3  WIP: avoid zip and unzip to improve the speed ...  ...   
4                       CLN: Deprecate convert_float  ...   

                                           milestone comments  \
0                                               None        0   
1                                               None        0   
2  {'url': 'https://api.github.com/repos/pandas-d...        0   
3                                               None        0   
4                                               None        2   

                 created_at                updated_at closed_at  \
0 2021-04-27 14:57:36+00:00 2021-04-27 15:05:47+00:00       NaT   
1 2021-04-27 14:34:14+00:00 2021-04-27 14:54:49+00:00       NaT   
2 2021-04-27 08:45:35+00:00 2021-04-27 12:59:49+00:00       NaT   
3 2021-04-27 07:46:58+00:00 2021-04-27 08:20:07+00:00       NaT   
4 2021-04-27 04:08:55+00:00 2021-04-27 13:57:21+00:00       NaT   

  author_association active_lock_reason  \
0               NONE                NaN   
1        CONTRIBUTOR                NaN   
2             MEMBER                NaN   
3               NONE                NaN   
4        CONTRIBUTOR                NaN   

                                                body performed_via_github_app  \
0  #### https://pandas.pydata.org/docs/reference/...                      NaN   
1                         clean-up of typing issues.                      NaN   
2  Follow-up on https://github.co

# XML and HTML Web Scrapping

Now, let us look at how we can create a dataframe from HTML websites and their urls

As we already discuused above, when we use the read_html() function, the result is a list of different tables used in the website.

Inorder to access different dataframes, we can use the list indexing methods which are demonstrated below

In [209]:
fail = pd.read_html(r'https://www.fdic.gov/resources/resolutions/bank-failures/failed-bank-list/')
failures = fail[0]
failures.head()

Bank Name               City State   Cert  \
0                 Almena State Bank             Almena    KS  15426   
1        First City Bank of Florida  Fort Walton Beach    FL  16748   
2              The First State Bank      Barboursville    WV  14361   
3                Ericson State Bank            Ericson    NE  18265   
4  City National Bank of New Jersey             Newark    NJ  21111   

        Acquiring Institution       Closing Date  
0                 Equity Bank   October 23, 2020  
1   United Fidelity Bank, fsb   October 16, 2020  
2              MVB Bank, Inc.      April 3, 2020  
3  Farmers and Merchants Bank  February 14, 2020  
4             Industrial Bank   November 1, 2019

In [210]:
failures

Bank Name               City State   Cert  \
0                    Almena State Bank             Almena    KS  15426   
1           First City Bank of Florida  Fort Walton Beach    FL  16748   
2                 The First State Bank      Barboursville    WV  14361   
3                   Ericson State Bank            Ericson    NE  18265   
4     City National Bank of New Jersey             Newark    NJ  21111   
..                                 ...                ...   ...    ...   
558                 Superior Bank, FSB           Hinsdale    IL  32646   
559                Malta National Bank              Malta    OH   6629   
560    First Alliance Bank & Trust Co.         Manchester    NH  34264   
561  National State Bank of Metropolis         Metropolis    IL   3815   
562                   Bank of Honolulu           Honolulu    HI  21029   

                   Acquiring Institution       Closing Date  
0                            Equity Bank   October 23, 2020  
1              United Fidelity Bank, fsb   October 16, 2020  
2                         MVB Bank, Inc.      April 3, 2020  
3             Farmers and Merchants Bank  February 14, 2020  
4                        Industrial Bank   November 1, 2019  
..                                   ...                ...  
558                Superior Federal, FSB      July 27, 2001  
559                    North Valley Bank        May 3, 2001  
560  Southern New Hampshire Bank & Trust   February 2, 2001  
561              Banterra Bank of Marion  December 14, 2000  
562                   Bank of the Orient   October 13, 2000  

[563 rows x 6 columns]

## Binary Data Formats

Now let us look at the way in which we can write the dataset from a csv format to a byte format which isn't readable to the human eye.

This kind of data is called a 'Pickle File' and is important for deploying our models

In [211]:
frame = pd.read_csv('pokemon_data.csv')
frame.to_pickle('Pickle_File')

#Write  Operation

The to_pickle() function is used to write the csv file into a pickle format file in a specified location and name.

The operation of writing the data or converting it from object to byte code  is "Serialisation" or "Write operation"

When reading the data from byte code and creating a Dataframe, it is "Deserialisation" or "Write Opearion"

The combination of these both is called the "Serde" operation which means Serialisation/Deserialisation

In [212]:
#Read function
pd.read_pickle('Pickle_File')

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

# Web APIs

An API stands for "Application Programming Interface" 

It acts as a connection between two homogeneous or heterogeneous systems or databases.

APIS work on the principle of "Request and Response". Let us understand it in a Leymann way.

When we try to withdraw money from an ATM, the ATM first sends a request to the bank server to check if the required amount of money is available in the account. This is called 'Request'.

Once it recieves a confirmation that money is available, it sends a response to the ATM to start giving out the money. This is called a 'Response'.


An API acts as a bridge between these two systems ATM and Bank Server and integrates the functionalities of both

In [149]:
import requests
url= 'https://api.github.com/repos/pandas-dev/pandas/issues'
resp = requests.get(url)
resp

#request and response way.

<Response [200]>

We use a default library called 'requests' available in python to handle responses and requests.

In the above code, we are trying to specify the url and requesting the data from the url and storing it in an object called resp.

The output resp is given as <Response [200]>. Here 200 indicates a successful response to the request

In [152]:
data = resp.json()
data

[{'url': 'https://api.github.com/repos/pandas-dev/pandas/issues/41178',
  'repository_url': 'https://api.github.com/repos/pandas-dev/pandas',
  'labels_url': 'https://api.github.com/repos/pandas-dev/pandas/issues/41178/labels{/name}',
  'comments_url': 'https://api.github.com/repos/pandas-dev/pandas/issues/41178/comments',
  'events_url': 'https://api.github.com/repos/pandas-dev/pandas/issues/41178/events',
  'html_url': 'https://github.com/pandas-dev/pandas/pull/41178',
  'id': 868617307,
  'node_id': 'MDExOlB1bGxSZXF1ZXN0NjI0MTEzNTIw',
  'number': 41178,
  'title': 'TST: run frame/series arithmetic tests with+without numexpr',
  'user': {'login': 'jorisvandenbossche',
   'id': 1020496,
   'node_id': 'MDQ6VXNlcjEwMjA0OTY=',
   'avatar_url': 'https://avatars.githubusercontent.com/u/1020496?v=4',
   'gravatar_id': '',
   'url': 'https://api.github.com/users/jorisvandenbossche',
   'html_url': 'https://github.com/jorisvandenbossche',
   'followers_url': 'https://api.github.com/users/jori

Now that we obtained the response successfully, we proceed to read the data.

Since the data is avaialable in JSON, we use the .json() function to read the data.

Now, that we have read the data, we can perform modifications and create dataframes from it just like we did it in the JSON section

In [213]:
data[0]['title']

'TST: run frame/series arithmetic tests with+without numexpr'

In [214]:
data[0]['user']['id']

1020496

In [215]:
for i in range(len(data)):
    print(data[i]['user']['id'])

1020496
1304505
50434302
8078968
3578739
82109078
14043325
25440862
50263213
1020496
50434302
8078968
1020496
1020496
10372220
1020496
41898282
10372220
24256554
47963215
48971969
26364415
13501084
34571894
8078968
61934744
5946712
24256554
24256554
45742869


In [216]:
pd.DataFrame(data,columns=['number','title','user','labels'])

number                                              title  \
0    41178  TST: run frame/series arithmetic tests with+wi...   
1    41177  WIP: avoid zip and unzip to improve the speed ...   
2    41176                       CLN: Deprecate convert_float   
3    41175                 PERF: implement NDArrayBackedBlock   
4    41174  BUG: Unexpected output of `to_json` for `compl...   
..     ...                                                ...   
25   41146  BUG: raise ValueError when sep and delimiter a...   
26   41145  BUG: isin fails to correctly compare None values.   
27   41143      ENH: Make Styler work with non-unique indexes   
28   41142  API: refactor Styler to allow sparse cols or i...   
29   41141                  BUG: Fix series clipping NA issue   

                                                 user  \
0   {'login': 'jorisvandenbossche', 'id': 1020496,...   
1   {'login': 'qiuwei', 'id': 1304505, 'node_id': ...   
2   {'login': 'ahawryluk', 'id': 50434302, 'node_i...   
3   {'login': 'jbrockmendel', 'id': 8078968, 'node...   
4   {'login': 'karalekas', 'id': 3578739, 'node_id...   
..                                                ...   
25  {'login': 'phofl', 'id': 61934744, 'node_id': ...   
26  {'login': 'DriesSchaumont', 'id': 5946712, 'no...   
27  {'login': 'attack68', 'id': 24256554, 'node_id...   
28  {'login': 'attack68', 'id': 24256554, 'node_id...   
29  {'login': 'Rasori', 'id': 45742869, 'node_id':...   

                                               labels  
0   [{'id': 47223669, 'node_id': 'MDU6TGFiZWw0NzIy...  
1                                                  []  
2                                                  []  
3                                                  []  
4   [{'id': 76811, 'node_id': 'MDU6TGFiZWw3NjgxMQ=...  
..                                                ...  
25  [{'id': 42670965, 'node_id': 'MDU6TGFiZWw0MjY3...  
26  [{'id': 57296398, 'node_id': 'MDU6TGFiZWw1NzI5...  
27  [{'id': 76812, 'node_id': 'MDU6TGFiZWw3NjgxMg=...  
28  [{'id': 1741841389, 'node_id': 'MDU6TGFiZWwxNz...  
29  [{'id': 57296398, 'node_id': 'MDU6TGFiZWw1NzI5...  

[30 rows x 4 columns]